File: src/levels/L17-pedidos-en-vivo/ts/events_command.ts
// Cada evento del pedido es un comando: se encola y se puede retirar antes de ejecutarse.
class OrderEventCommand {
  constructor(private order: Order, private event: string, readonly id: number) {}
  execute() { this.order.handle(this.event) }
}

class FrontDesk {
  constructor(private queue: CommandQueue, private order: Order) {}

  // region: FrontDesk#submit
  submit(request: EventRequest) {
    if (request.undo) return this.queue.cancel(request.commandId)
    this.queue.push(new OrderEventCommand(this.order, request.event, request.commandId))
  }
  // endregion
}

class CommandQueue {
  private pending: OrderEventCommand[] = []

  // region: CommandQueue#push
  push(command: OrderEventCommand) { this.pending.push(command) }
  cancel(commandId: number) { this.pending = this.pending.filter((c) => c.id !== commandId) }
  runNext() { this.pending.shift()?.execute() }
  // endregion
}

File: src/levels/L17-pedidos-en-vivo/events_command.rb
# Cada evento del pedido es un comando: se encola y se puede retirar antes de ejecutarse.
class OrderEventCommand
  attr_reader :id

  def initialize(order, event, id) = (@order, @event, @id = order, event, id)
  def execute = @order.handle(@event)
end

class FrontDesk
  # region: FrontDesk#submit
  def submit(request)
    return @queue.cancel(request.command_id) if request.undo?
    @queue.push(OrderEventCommand.new(@order, request.event, request.command_id))
  end
  # endregion
end

class CommandQueue
  # region: CommandQueue#push
  def push(command) = @pending << command
  def cancel(command_id) = @pending.reject! { |c| c.id == command_id }
  def run_next = @pending.shift&.execute
  # endregion
end
